parquet-column/src/main/java/parquet/schema/Types.java

```diff
@@ -555,6 +622,1024 @@ protected GroupType build(String name) {
+      }
+
+      public P named(String name) {
+        parent.setValueType(build("value"));
+        return parent.named(name);
+      }
+
+      @Override
+      protected MapGroupValueBuilder<P, Q> self() {
+        return this;
+      }
+    }
+
+    public static class MapMapValueBuilder<P, Q extends BaseMapBuilder<P, Q>> extends
+        BaseMapBuilder<P, MapMapValueBuilder<P, Q>> {
+      private final BaseMapBuilder<P, Q> parent;
+
+      public MapMapValueBuilder(BaseMapBuilder<P, Q> parent) {
+        super(parent.parent);
+        this.parent = parent;
+      }
+
+      public P named(String name) {
+        parent.setValueType(build("value"));
+        return parent.named(name);
+      }
+
+      @Override
+      protected MapMapValueBuilder<P, Q> self() {
+        return this;
+      }
+    }
+
+    public static class MapListValueBuilder<P, Q extends BaseMapBuilder<P, Q>> extends
+        BaseListBuilder<P, MapListValueBuilder<P, Q>> {
+      private final BaseMapBuilder<P, Q> parent;
+
+      public MapListValueBuilder(BaseMapBuilder<P, Q> parent) {
+        super(parent.parent);
+        this.parent = parent;
+      }
+
+      public P named(String name) {
+        parent.setValueType(build("value"));
+        return parent.named(name);
+      }
+
+      @Override
+      protected MapListValueBuilder<P, Q> self() {
+        return this;
+      }
+    }
+
+    public static class MapListKeyBuilder<P, Q extends BaseMapBuilder<P, Q>> extends
+        BaseListBuilder<P, MapListKeyBuilder<P, Q>> {
+      private final BaseMapBuilder<P, Q> parent;
+
+      public MapListKeyBuilder(BaseMapBuilder<P, Q> parent) {
+        super(parent.parent);
+        this.parent = parent;
+      }
+
+      public P named(String name) {
+        return super.named(name);
+      }
+
+      @Override
+      protected MapListKeyBuilder<P, Q> self() {
+        return this;
+      }
+
+      public void setKey(Type key) {
+        parent.setKeyType(key);
+      }
+    }
+
+    protected void setKeyType(Type keyType) {
+      this.keyType = keyType;
+    }
+
+    protected void setValueType(Type valueType) {
+      this.valueType = valueType;
+    }
+
+    private Type keyType = null;
+    private Type valueType = null;
+    private Class<P> returnType = null;
+
+    public BaseMapBuilder(P parent) {
+      super(parent);
+    }
+
+    private BaseMapBuilder(Class<P> returnType) {
+      super(returnType);
+      this.returnType = returnType;
+    }
+
+    @Override
+    protected abstract Q self();
+
+    public MapKeyBuilder<P, Q> key(PrimitiveTypeName type) {
+      if (parent != null) {
+        return new MapKeyBuilder<P, Q>(this, type);
+      } else {
+        return new MapKeyBuilder<P, Q>(this, returnType, type);
+      }
+    }
+
+    public MapGroupKeyBuilder<P, Q> groupKey() {
+      return new MapGroupKeyBuilder<P, Q>(this);
+    }
+
+    public MapListKeyBuilder<P, Q> listKey() {
+      return new MapListKeyBuilder<P, Q>(this);
+    }
+
```

Comment:
Okay, I think we still need a couple of methods that were overlooked: `key(Type)` so you can make more complex types than are supported by the builders. Then we will also need `value(Type)`, `requiredValue(Type)`, and `optionalValue(Type)` on the key builders that set the value type.